### PR TITLE
Gas defaults

### PR DESCRIPTION
**Pull request by ethanbennett**

### CODE REVIEW

File: test/eth/TransactionManager.spec.js [hunk 1]
```diff
@@ -158,14 +157,6 @@ describe('lifecycle hooks', () => {
     expect(lockHandlers.initialized).toBeCalledTimes(5);
     expect(lockHandlers.pending).toBeCalledTimes(5);
     expect(lockHandlers.mined).toBeCalledTimes(5);
-
-    log('\ndraw');
```
Comment (levity) [inline test/eth/TransactionManager.spec.js]: why remove these?

File: src/eth/EthereumCdpService.js [hunk 2]
```diff
@@ -155,7 +158,7 @@ export default class EthereumCdpService extends PrivateService {
       unit,
       promise
     });
-    if (waitForConfirm) await this._txMgr().confirm(convert);
```
Comment (levity) [inline src/eth/EthereumCdpService.js]: oh crap, my bad

File: src/eth/GasEstimatorService.js [hunk 3]
```diff
@@ -3,32 +3,42 @@ import { PublicService } from '@makerdao/services-core';
 export default class GasEstimatorService extends PublicService {
   constructor(name = 'gasEstimator') {
     super(name, ['web3', 'log']);
-    this._percentage = null;
+    this._percentage = 1.55;
```
Comment (levity) [inline src/eth/GasEstimatorService.js]: calling this `multiplier` instead of `percentage` makes more sense to me

File: src/eth/GasEstimatorService.js [hunk 4]
```diff
@@ -3,32 +3,42 @@ import { PublicService } from '@makerdao/services-core';
...
-      if (this._percentage === null && this._absolute !== null) {
-        return Math.min(this._absolute, blockLimit);
-      }
+    const blockLimit = web3Data[0].gasLimit;
+    const estimate = web3Data[1];
 
-      if (this._absolute === null) {
-        return Math.min(estimate * this._percentage, blockLimit);
-      }
+    if (this._percentage === null && this._absolute !== null) {
```
Comment (levity) [inline src/eth/GasEstimatorService.js]: if you know that 0 is not a valid value for these variables, then there's no need to explicitly check equality with null -- just check truthiness or falsiness (i.e. `if (!this._percentage && this._absolute)`)

File: src/eth/GasEstimatorService.js [hunk 5]
```diff
@@ -3,32 +3,42 @@ import { PublicService } from '@makerdao/services-core';
...
+        blockLimit
+      );
+    }
 
-      return Math.min(estimate * this._percentage, this._absolute, blockLimit);
-    });
+    return options;
   }
 
   setPercentage(number) {
```
Comment (levity) [inline src/eth/GasEstimatorService.js]: this is a good opportunity to use [getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) and [setters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set)

File: test/Maker.spec.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ async function createMaker(privateKey) {
 }
 
 test('create without any options', async () => {
-  await Maker.create('test');
+  await Maker.create('test', { log: false });
```
Comment (levity) [inline test/Maker.spec.js]: please remove this -- the test is specifically checking `create` without a second argument

File: test/eth/TransactionObject.spec.js [hunk 7]
```diff
@@ -96,7 +96,11 @@ describe('normal web service behavior', () => {
 
     test(
       'generic error',
-      testErrorHandling(() => mkr.transfer(testAddress, '2000000'), /reverted/)
+      testErrorHandling(
+        () => mkr.transfer(testAddress, '2000000'),
+        /reverted/,
+        false
```
Comment (levity) [inline test/eth/TransactionObject.spec.js]: why was this added?

File: src/eth/TransactionManager.js [hunk 8]
```diff
@@ -57,7 +57,20 @@ export default class TransactionManager extends PublicService {
       (async () => {
         // so we do our async operations inside this immediately-executed
         // async function.
-        const txOptions = await this._buildTransactionOptions(options);
+        let data = contract.interface.functions[method](...args).data;
+        if (options.dsProxy) {
+          const proxyAddress = await this.get('proxy').currentProxy();
+          const proxy = this.get('proxy')._getUnwrappedProxyContract(
```
Comment (levity) [inline src/eth/TransactionManager.js]: if `_getUnwrappedProxyContract` is now expected to be used outside the service it shouldn't start with `_`

File: src/eth/TransactionManager.js [hunk 9]
```diff
@@ -57,7 +57,20 @@ export default class TransactionManager extends PublicService {
       (async () => {
         // so we do our async operations inside this immediately-executed
         // async function.
-        const txOptions = await this._buildTransactionOptions(options);
+        let data = contract.interface.functions[method](...args).data;
```
Comment (levity) [inline src/eth/TransactionManager.js]: it would be cleaner to change this to 
```js
const txOptions = await this._buildTransactionOptions(options, contract);
```
and then have all the rest of the changes inside `_buildTransactionOptions`

File: test/eth/TransactionManager.spec.js (hunk 1 above)
Comment (ethanbennett) [reply test/eth/TransactionManager.spec.js]: From what I can tell, they weren't actually being used for anything. It was timing out before it got through `wipe`, so I could put them back and increase the timeout on this test, but I think they might as well be removed if they're not testing anything.

File: src/eth/GasEstimatorService.js (hunk 4 above)
Comment (ethanbennett) [reply src/eth/GasEstimatorService.js]: Good call--this and the get/set functions were already in this service, they just weren't working

File: test/eth/TransactionObject.spec.js (hunk 7 above)
Comment (ethanbennett) [reply test/eth/TransactionObject.spec.js]: I think I can actually remove it now. Initially, web3's gas estimation function would go ahead and revert the transaction before it actually tried to send it, which meant that it was never pending. It caused unpredictable behavior in other error handling scenarios though, so I put `estimateGas` in a try/catch (and assigned the fallback gas limit in the catch) so transactions now fail where the tx manager expects them to.

File: test/eth/TransactionManager.spec.js (hunk 1 above)
Comment (levity) [reply test/eth/TransactionManager.spec.js]: OK, fair enough :+1:

File: src/eth/EthereumCdpService.js (hunk 2 above)
Comment (levity) [reply src/eth/EthereumCdpService.js]: pulled this fix into master (ad04613)

File: test/eth/GasEstimatorService.spec.js [hunk 10]
```diff
@@ -17,142 +22,52 @@ function buildTestGasEstimatorService() {
...
-
-  gasEstimator
-    .manager()
-    .connect()
-    .then(() => {
-      expect(gasEstimator.getPercentage()).toBe(null);
-      expect(gasEstimator.getAbsolute()).toBe(null);
-      done();
-    });
+test('policies initially null', () => {
```
Comment (tyler17) [inline test/eth/GasEstimatorService.spec.js]: super minor suggestion: might as well check the fallback value too.  Can also change the name of the test

File: lib/dai-plugin-mcd/test/helpers.js [hunk 11]
```diff
@@ -63,7 +63,7 @@ export async function setDebtCeiling(maker, amount, ilk) {
       ilk && stringToBytes(ilk),
       stringToBytes(ilk ? 'line' : 'Line'),
       ethAbi.encodeParameter('uint', amount.toFixed('rad')),
-      { dsProxy: scs.getContractAddress('MCD_MOM') }
+      { dsProxy: scs.getContractAddress('MCD_MOM'), gasLimit: 4000000 }
```
Comment (tyler17) [inline lib/dai-plugin-mcd/test/helpers.js]: is this necessary? I tried deleting it and tests still passed

Comment (ethanbennett) [reply lib/dai-plugin-mcd/test/helpers.js]: I think it's not necessary anymore now that I added the fallback gas limit to the estimator service. Good catch 👍

File: src/eth/GasEstimatorService.js [hunk 12]
```diff
@@ -3,62 +3,88 @@ import { PublicService } from '@makerdao/services-core';
...
-    this._percentage = null;
+    this._multiplier = 1.55;
     this._absolute = null;
+    this._fallback = 4000000;
   }
 
-  estimateGasLimit(transaction) {
-    if (this._percentage === null && this._absolute === null) {
-      throw new Error('no gas limit policy set');
+  async estimateGasLimit(transaction, options = {}) {
```
Comment (levity) [inline src/eth/GasEstimatorService.js]: it would be better here just to return the bare value rather than mutating options

File: src/eth/TransactionManager.js [hunk 13]
```diff
@@ -149,12 +154,47 @@ export default class TransactionManager extends PublicService {
...
+
+    transaction = {
+      from: this.get('web3').currentAddress(),
+      nonce: nonce,
+      to: options.dsProxy ? proxyAddress : contract.address,
+      data: data,
+      ...transaction
     };
+
+    return await this.get('gasEstimator').estimateGasLimit(
```
Comment (levity) [inline src/eth/TransactionManager.js]: no need for `return await`

File: src/eth/TransactionManager.js [hunk 14]
```diff
@@ -149,12 +154,47 @@ export default class TransactionManager extends PublicService {
     return minePromise;
   }
 
-  async _buildTransactionOptions(data) {
+  async _buildTransactionOptions(options, contract, method, args) {
+    const nonce = await this.get('nonce').getNonce();
+
+    if (contract && !options.gasLimit) {
+      await this._getGasLimit(options, contract, method, args, nonce);
```
Comment (levity) [inline src/eth/TransactionManager.js]: likewise here, if instead of changing things via hidden side-effects, it were just `options.gasLimit = await this._getGasLimit(...)` it would be easier to understand what's happening

Comment (levity) [reply src/eth/TransactionManager.js]: also i don't think you need to pass `nonce` because it's never read? nonce-handling could probably stay as it was

File: src/eth/TransactionManager.js [hunk 15]
```diff
@@ -149,12 +154,47 @@ export default class TransactionManager extends PublicService {
...
+
+    if (options.value) {
+      transaction.value = options.value;
+    }
+
+    transaction = {
+      from: this.get('web3').currentAddress(),
+      nonce: nonce,
+      to: options.dsProxy ? proxyAddress : contract.address,
+      data: data,
```
Comment (levity) [inline src/eth/TransactionManager.js]: reminder that you can use the shorthand syntax whenever the key and variable name are the same

File: src/eth/TransactionManager.js (hunk 14 above)
Comment (ethanbennett) [reply src/eth/TransactionManager.js]: The nonce actually needs to be passed into `estimateGas`, but if I get the nonce in `_getGasLimit` there would be some scenarios where it's left out from the final options

File: test/plugins.spec.js [hunk 16]
```diff
@@ -265,9 +265,8 @@ test('each plugin type can be created with an optional options object for additi
 
   const functionPlugin = jest.fn((maker, config, options) => {
     const gasEstimatorService = maker.service('gasEstimator', true);
-    gasEstimatorService.setPercentage(options.percentage);
-    const percentage = gasEstimatorService.getPercentage();
-    expect(percentage).toBe(options.percentage);
+    gasEstimatorService.percentage = options.percentage;
```
Comment (levity) [inline test/plugins.spec.js]: oh wait, one last thing -- this is still referring to `percentage`
